### PR TITLE
Avoid buffering to tempfile when pushing with V2

### LINKED ISSUE #13309: Docker pull progressively slow for large layers
I am pulling an image from a private docker repo. It consist from multiple layers. I can see how small layers (as example 20MB) downloaded in 5-6 second (~4Mb/s).

However, with bigger layers (as example 300MB) it works fast at the beginning, but starts to slow down at the end and gets to ~0.1Mb/s (40 times slower than smaller layers or beginning of this layer)

### PR DESCRIPTION
**Pull request by stevvooe**
The practice of buffering to a tempfile during a pushing contributes massively
to slow V2 push performance perception. The protocol was actually designed to
avoid precalculation, supporting cut-through data push. This means we can
assemble the layer, calculate its digest and push to the remote endpoint, all
at the same time.

This should increase performance massively on systems with slow disks or IO
bottlenecks.

Closes #13309

Signed-off-by: Stephen J Day stephen.day@docker.com

### CODE REVIEW

File: graph/push_v2.go [hunk 1]
```diff
@@ -212,52 +212,46 @@ func (p *v2Pusher) pushV2Image(bs distribution.BlobService, img *image.Image) (d
...
 		Out:       out,
 		Formatter: p.sf,
-		Size:      size,
-		NewLines:  false,
-		ID:        stringid.TruncateID(img.ID),
-		Action:    "Pushing",
+
+		// TODO(stevvooe): This may cause a size reporting error. Try to get
+		// this from tar-split or elsewhere. The main issue here is that we
+		// don't want to buffer to disk *just* to calculate the size.
```
Comment (tiborvass) [inline graph/push_v2.go]: Ping @vbatts suggestions?

File: graph/push_v2.go [hunk 2]
```diff
@@ -212,52 +212,46 @@ func (p *v2Pusher) pushV2Image(bs distribution.BlobService, img *image.Image) (d
...
-	defer func() {
-		tf.Close()
-		os.Remove(tf.Name())
-	}()
-
-	size, dgst, err := bufferToFile(tf, arch)
-	if err != nil {
-		return "", err
-	}
+	defer arch.Close()
```
Comment (aaronlehmann) [inline graph/push_v2.go]: Where was arch closed before this commit?

File: graph/push_v2.go (hunk 1 above)
Comment (vbatts) [inline graph/push_v2.go]: interesting. With tar-split, we could plumb to get the sum size of the uncompressed tar archive, but that would not be accurate for the compressed blob.

File: graph/push_v2.go (hunk 2 above)
Comment (stevvooe) [inline graph/push_v2.go]: I don't think it was being closed.

File: graph/push_v2.go (hunk 1 above)
Comment (stevvooe) [inline graph/push_v2.go]: We can probably write the size somewhere.

I'd rather us take that as a regression and get this fix, rather than save correct size reporting at the cost of bad performance.

Comment (dmcgowan) [inline graph/push_v2.go]: This really was ugly before with the incorrect size. Progress reader could be updated to only use the size as an estimate instead, it has already been updated to not overrun the progress bar.

Current
```
428b411c28f0: Pushing [==================================================>] 197.2 MB/188.1 MB
```

Maybe should just be
```
428b411c28f0: Pushing [==================================================>] 197.2 MB
```

Comment (stevvooe) [inline graph/push_v2.go]: @dmcgowan Let's take that as a regression. The performance issue is awful.

Comment (stevvooe) [inline graph/push_v2.go]: @dmcgowan Or, where do I make that change?

Comment (dmcgowan) [inline graph/push_v2.go]: https://github.com/docker/docker/blob/master/pkg/jsonmessage/jsonmessage.go#L70

Comment (stevvooe) [inline graph/push_v2.go]: @dmcgowan That seems like a much larger change. Is that a condition of merging this fix?

We can probably get this information from a head request to the registry. This can be encapsulated in the blob statter. Would this be a better approach?

**Comment by ncdc** [comment]
FYI I tested this locally and it cut down a 2m10s push to 46s!!!

File: graph/push_v2.go (hunk 1 above)
Comment (dmcgowan) [inline graph/push_v2.go]: It just looks bad, like we can't do ~~math~~ arithmetic :stuck_out_tongue_winking_eye:

Comment (dmcgowan) [inline graph/push_v2.go]:
```diff
diff --git a/pkg/jsonmessage/jsonmessage.go b/pkg/jsonmessage/jsonmessage.go
index f3bcc8e..b4165da 100644
--- a/pkg/jsonmessage/jsonmessage.go
+++ b/pkg/jsonmessage/jsonmessage.go
@@ -67,7 +67,11 @@ func (p *JSONProgress) String() string {
                }
                pbBox = fmt.Sprintf("[%s>%s] ", strings.Repeat("=", percentage), strings.Repeat(" ", numSpaces))
        }
-       numbersBox = fmt.Sprintf("%8v/%v", current, total)
+       if p.Current > p.Total {
+               numbersBox = fmt.Sprintf("%8v", current)
+       } else {
+               numbersBox = fmt.Sprintf("%8v/%v", current, total)
+       }
 
        if p.Current > 0 && p.Start > 0 && percentage < 50 {
                fromStart := time.Now().UTC().Sub(time.Unix(p.Start, 0))

```

Comment (thaJeztah) [inline graph/push_v2.go]: Think that warrants a FIXME comment :wink:

**Comment by LK4D4** [comment]
Test failed :/

**Comment by jfrazelle** [comment]
please rebase

File: graph/push_v2.go (hunk 1 above)
Comment (stevvooe) [inline graph/push_v2.go]: This comment string is actually on the TODO.

File: graph/push_v2.go [hunk 3]
```diff
@@ -212,52 +212,46 @@ func (p *v2Pusher) pushV2Image(bs distribution.BlobService, img *image.Image) (d
...
-	desc := distribution.Descriptor{Digest: dgst}
-	if _, err := layerUpload.Commit(nil, desc); err != nil {
+	dgst := digester.Digest()
+	if _, err := layerUpload.Commit(context.Background(), distribution.Descriptor{Digest: dgst}); err != nil {
 		return "", err
 	}
 
-	out.Write(p.sf.FormatProgress(stringid.TruncateID(img.ID), "Image successfully pushed", nil))
+	logrus.Debugf("uploaded layer %s (%s), %d bytes", img.ID, dgst, nn)
+	out.Write(p.sf.FormatProgress(stringid.TruncateID(img.ID), "Pushed", nil))
```
Comment (dmcgowan) [inline graph/push_v2.go]: Why the change from `Image successfully pushed` to `Pushed`?

Feels like the sentence was cutoff when reading but perhaps I am just not used to it
```
The push refers to a repository [localhost:5000/ubuntu-1] (len: 1)
6d4946999d4f: Pushed 
9fd3c8c9af32: Pushed 
435050075b3f: Pushed 
428b411c28f0: Pushed 
latest: digest: sha256:02d84269c2fb70fbef7457da4f1a4ae44bab2ab02bfe494620d1076a5e1a5d0c size: 7720
```

Comment (stevvooe) [inline graph/push_v2.go]: @dmcgowan This is more in line with the messages around it. It is also an inaccurate statement. An image isn't being pushed. It is a layer.

**Comment by stevvooe** [comment]
@jfrazelle Rebased.

Failed tests on first run. Not sure if related.

@ncdc Thank you for the validation!

File: graph/push_v2.go (hunk 1 above)
Comment (thaJeztah) [inline graph/push_v2.go]: Oh! My bad, looked at the diff above, should've scrolled up.

**Comment by stevvooe** [comment]
The unit test is dependent on a data race between the two pushes (see https://github.com/docker/docker/blob/master/integration-cli/docker_cli_push_test.go#L117). Because the push faster, the unit test misses the race. I was able to produce the correct result locally with this branch using the hub as a target:

I had this in one shell:

```console
# ./bundles/1.9.0-dev/binary/docker push stevvooe/busybox
The push refers to a repository [docker.io/stevvooe/busybox] (len: 1)
8c2e06607696: Pushed
6ce2e90b0bc7: Pushed
cf2616975b4a: Pushed
latest: digest: sha256:45ff8f9c2ab391d1f179d0d62c0ee33966e31687e225689f29c952878ca016a3 size: 5061
```

Ran this simultaneously in the other shell:

```console
# ./bundles/1.9.0-dev/binary/docker push stevvooe/busybox
The push refers to a repository [docker.io/stevvooe/busybox] (len: 1)
push stevvooe/busybox is already in progress
```

I've decreased the wait time by 100ms to see if we can catch the race.

**Comment by stevvooe** [comment]
Reduced to 10ms to leverage race condition... worked locally.

**Comment by stevvooe** [comment]
Tried 20ms failed, trying 50ms.

Why do we have such silly tests?

**Comment by stevvooe** [comment]
Looks like `DockerRegistrySuite.TestPushInterrupt` is now passing. Seems like `DockerDaemonSuite.TestCleanupMountsAfterCrash` is failing, but that likely isn't related.

**Comment by stevvooe** [comment]
@jfrazelle Rebased and unit test fixed. There is still an unrelated failure.

**Comment by dmcgowan** [comment]
I don't know why that test will not pass on janky, tried restarting multiple times. It succeeds locally so...

LGTM

**Comment by LK4D4** [comment]
Yeah, I think janky test is flaky.

**Comment by RichardScothern** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM

**Comment by thaJeztah** [comment]
ping @calavera for cherry-pick

**Comment by marcogallotta** [comment]
@calavera looks like this is in the changelog for 1.8.2 but not cherry-picked? We're pretty eager to see this get in.

**Comment by stevvooe** [comment]
@calavera cherry-pick?

**Comment by calavera** [comment]
Sorry, I missed to cherry picked this change and it went to the changelog when it shouldn't have gone.

After further consideration, I've decided to leave this change for 1.9. I don't think we should consider performance perception a critical bug to fix in a patch release, and 1.9 will be released it soon anyways.

**Comment by stevvooe** [comment]
@calavera This buffering penalty is causing docker to be unusable in certain use cases. In particular, this will effect small instances with low disk buffer cache and slow disks or any use cases that have large images.

**Comment by lwcolton** [comment]
@calavera I disagree that this is not a critical issue.  I spent quite a bit of time upgrading to registry v2 in an attempt to gain better performance, and instead have made `docker push` unusable.

**Comment by lwcolton** [comment]
To be as helpful as possible:  I got here from #13250 and believe this is the issue I'm seeing.  High CPU and I/O wait during a painfully long "Buffering to disk" phase.  If there's anything else I can do to help let me know.

**Comment by stevvooe** [comment]
@lwcolton I agree with you. This is indeed a critical issue that is impacting users. This patch virtually eliminates the buffering to disk phase, since there isn't really buffering to disk.

@calavera I'm a little disappointed this didn't get in. While I agree that performance optimizations are not critical issues, a performance defect actually is. If you examine the sheer number of users impacted by this problem, it is clear that this belongs in a patch release.

**Comment by vik2017** [comment]
Hey anyone please let me know in what docker version I can see this changes because I'm using Version: 1.11.2, API version: 1.23, but still facing the same issue

**Comment by LK4D4** [comment]
@vik2017 it was in 1.9. Please note that slow push/pull can be because of many things(like slow gzip or unstable networking).

**Comment by vik2017** [comment]
thanks Alexander :) , I will have a look into that, if those are the things causing problems

**Comment by vik2017** [comment]
Hey ALexander I'm new to Docker and trying to fix the performance issue of push command. I'm wondering what are the steps in the docker push command(by this I mean, what actually happens in push command). I tried to look at the docs but didn't get much info. I want to know this so that I can figure out where gzip is involved and to get more understanding about the push.

**Comment by stevvooe** [comment]
@vik2017 Also, I'd recommend opening another issue and explaining your _exact_ situation. It is unlikely to be related to the original issue that started this PR.

**Comment by LK4D4** [comment]
@vik2017 we have #17674 for improving compressing speed.

**Comment by vik2017** [comment]
Thanks Stephen. I have started looking at the code now. I'm also gonna check if there are other factors behind the issue.  I would give myself a day or two to figure out before I open other issue just to make sure everything is alright on my side :p

**Comment by vik2017** [comment]
Thanks Alexander. I just saw that the API version in my system is  1.23 , so does that mean it's v1 and not v2?
